### PR TITLE
Olofk/pcacherace

### PR DESCRIPTION
**Pull request by olofkallander**
SsrcOutboundContext and packetcache caused multiple races.
Fixed by moving some operations onto Transport Jobs and some shared items made atomic.

### CODE REVIEW

File: bridge/engine/RecordingVideoForwarderSendJob.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+void RecordingVideoForwarderSendJob::run()
+{
+    auto rtpHeader = rtp::RtpHeader::fromPacket(*_packet);
+    if (!rtpHeader)
+    {
+        assert(false); // should have been checked multiple times
+        return;
+    }
+
+    if (_outboundContext.rtpMap.format == RtpMap::Format::VP8RTX)
```
Comment (olofkallander) [inline bridge/engine/RecordingVideoForwarderSendJob.cpp:43]: Is this actually possible for recording streams?

File: bridge/engine/RecordingVideoForwarderSendJob.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+            logger::debug("%s requested key frame from %u on ssrc %u",
+                "RecordingVideoForwarderSendJob",
+                _transport.getLoggableId().c_str(),
+                _senderInboundContext.ssrc,
+                _outboundContext.ssrc);
+        }
+    }
+
+    if (!_outboundContext.shouldSend(rtpHeader->ssrc, _extendedSequenceNumber))
```
Comment (olofkallander) [inline bridge/engine/RecordingVideoForwarderSendJob.cpp:102]: This is where we check if the packet happens to be too old (512 pkts). It also always allows a new source ssrc to pass through as it means new source stream for rewrite. I guess for recording that rewrite is not actually happening.

File: bridge/engine/SsrcOutboundContext.h [hunk 3]
```diff
@@ -99,32 +95,22 @@ class SsrcOutboundContext
...
+    {
+        lastSendTime = timestamp;
+        idle = false;
+    }
 
     // Stream owner is being removed. Stop outbound packets over this context
     bool markedForDeletion;
-
     // Stream is temporarily idle and srtp encryption context is removed. It may be activated again.
     bool idle;
```
Comment (olofkallander) [inline bridge/engine/SsrcOutboundContext.h:113]: I removed the PLI metric that appears only in logs. It was a source for data race.

File: bridge/engine/RecordingAudioForwarderSendJob.cpp [hunk 4]
```diff
@@ -53,11 +60,25 @@ void RecordingAudioForwarderSendJob::run()
         {
             if (!packetCache->add(*_packet, nextSequenceNumber))
             {
-                logger::info("Dropping rec audio packet - cache issues...", "RecordingAudioForwarderSendJob");
-                return;
+                logger::warn("Failed to cache rec audio packet", "RecordingAudioForwarderSendJob");
```
Comment (olofkallander) [inline bridge/engine/RecordingAudioForwarderSendJob.cpp:63]: let the audio packet through also when packet cache is not accepting it.

Comment (RicardoMDomingues) [reply bridge/engine/RecordingAudioForwarderSendJob.cpp:63]: Yes, that make sense 👍

**Review by RicardoMDomingues — APPROVED** [review]

**Pull request merged**